src/Native/Runtime/arm/AllocFast.S

```diff
@@ -2,115 +2,470 @@
+
+LOCAL_LABEL(RhpNewFast_RarePath):
+        mov         r0, r4            // restore EEType
+        mov         r1, #0
+        EPILOG_POP  "{r4,lr}"
+        b           C_FUNC(RhpNewObject)
+
+LEAF_END RhpNewFast, _TEXT
 
 // Allocate non-array object with finalizer.
 //  r0 == EEType
 //
-//    LEAF_ENTRY RhpNewFinalizable, _TEXT
-//#ifdef _DEBUG
-//        GEN_ASSERT_FUNC RhpNewFinalizableName
-//#endif
-//    LEAF_END RhpNewFinalizable, _TEXT
-//
+LEAF_ENTRY RhpNewFinalizable, _TEXT
+        mov         r1, #GC_ALLOC_FINALIZE
+        b           C_FUNC(RhpNewObject)
+LEAF_END RhpNewFinalizable, _TEXT
+
 
 // Allocate non-array object.
 //  r0 == EEType
 //  r1 == alloc flags
-    NESTED_ENTRY RhpNewObject, _TEXT, NoHandler
-#ifdef _DEBUG
-        GEN_ASSERT_FUNC RhpNewObjectName
-#endif
-    NESTED_END RhpNewObject, _TEXT
+NESTED_ENTRY RhpNewObject, _TEXT, NoHandler
+
+        PUSH_COOP_PINVOKE_FRAME r3
+
+        // r0: EEType
+        // r1: alloc flags
+        // r3: transition frame
+
+        // Preserve the EEType in r5.
+        mov         r5, r0
+
+        ldr         r2, [r0, #OFFSETOF__EEType__m_uBaseSize]    // cbSize
+
+        // void* RhpGcAlloc(EEType *pEEType, UInt32 uFlags, UIntNative cbSize, void * pTransitionFrame)
+        blx         C_FUNC(RhpGcAlloc)
+
+        // Set the new object's EEType pointer on success.
+        cbz         r0, LOCAL_LABEL(NewOutOfMemory)
+        str         r5, [r0, #OFFSETOF__Object__m_pEEType]
+
+        // If the object is bigger than RH_LARGE_OBJECT_SIZE, we must publish it to the BGC
+        ldr         r1, [r5, #OFFSETOF__EEType__m_uBaseSize]
+        movw        r2, #(RH_LARGE_OBJECT_SIZE & 0xFFFF)
+        movt        r2, #(RH_LARGE_OBJECT_SIZE >> 16)
+        cmp         r1, r2
+        blo         LOCAL_LABEL(New_SkipPublish)
+
+        // r0: already contains object
+        // r1: already contains object size
+
+        bl          C_FUNC(RhpPublishObject)
+        // r0: function returned the passed-in object
+
+LOCAL_LABEL(New_SkipPublish):
+
+        POP_COOP_PINVOKE_FRAME
+        bx          lr
+
+LOCAL_LABEL(NewOutOfMemory):
+        // This is the OOM failure path. We're going to tail-call to a managed helper that will throw
+        // an out of memory exception that the caller of this allocator understands.
+
+        mov         r0, r5            // EEType pointer
+        mov         r1, #0            // Indicate that we should throw OOM.
+
+        POP_COOP_PINVOKE_FRAME
+
+        b           C_FUNC(RhExceptionHandling_FailedAllocation)
+
+NESTED_END RhpNewObject, _TEXT
 
 // Allocate one dimensional, zero based array (SZARRAY).
 //  r0 == EEType
 //  r1 == element count
-//
-//    LEAF_ENTRY RhpNewArray, _TEXT
-//#ifdef _DEBUG
-//        GEN_ASSERT_FUNC RhpNewArrayName
-//#endif
-//    LEAF_END RhpNewArray, _TEXT
-//
+LEAF_ENTRY RhpNewArray, _TEXT
+        PROLOG_PUSH "{r4,lr}"
+        // we want to limit the element count to the non-negative 32-bit int range
```

Comment:
Why do we need this check? The Windows ARM version handles the negative check in ArraySizeBig and the `0x7fffffff` check doesn't even exist there.